qiskit_experiments/experiment_data.py

```diff
@@ -12,130 +12,321 @@
+        if isinstance(index, (int, slice)):
+            return self._data[index]
+        if isinstance(index, str):
+            return [data for data in self._data if data.get("job_id") == index]
+        raise QiskitError(f"Invalid index type {type(index)}.")
+
+    def add_figure(
+        self,
+        figure: Union[str, bytes],
+        figure_name: Optional[str] = None,
+        overwrite: bool = False,
+    ) -> Tuple[str, int]:
+        """Save the experiment figure.
+
+        Args:
+            figure: Name of the figure file or figure data to upload.
+            figure_name: Name of the figure. If ``None``, use the figure file name, if
+                given, or a generated name.
+            overwrite: Whether to overwrite the figure if one already exists with
+                the same name.
+
+        Returns:
+            A tuple of the name and size of the saved figure. Returned size
+            is 0 if there is no experiment service to use.
+
+        Raises:
+            QiskitError: If the figure with the same name already exists,
+                         and `overwrite=True` is not specified.
+        """
+        if not figure_name:
+            if isinstance(figure, str):
+                figure_name = figure
+            else:
+                figure_name = f"figure_{self.experiment_id}_{len(self.figure_names)}"
+
+        existing_figure = figure_name in self._figure_names
+        if existing_figure and not overwrite:
+            raise QiskitError(
+                f"A figure with the name {figure_name} for this experiment "
+                f"already exists. Specify overwrite=True if you "
+                f"want to overwrite it."
+            )
+        out = [figure_name, 0]
+        self._figures[figure_name] = figure
+        self._figure_names.append(figure_name)
+        return out
+
+    def figure(
+        self, figure_name: Union[str, int], file_name: Optional[str] = None
+    ) -> Union[int, bytes]:
+        """Retrieve the specified experiment figure.
+
+        Args:
+            figure_name: Name of the figure or figure position.
+            file_name: Name of the local file to save the figure to. If ``None``,
+                the content of the figure is returned instead.
+
+        Returns:
+            The size of the figure if `file_name` is specified. Otherwise the
+            content of the figure in bytes.
+
+        Raises:
+            QiskitError: If the figure cannot be found.
+        """
+        if isinstance(figure_name, int):
+            figure_name = self._figure_names[figure_name]
+
+        figure_data = self._figures.get(figure_name, None)
+        if figure_data is not None:
+            if isinstance(figure_data, str):
+                with open(figure_data, "rb") as file:
+                    figure_data = file.read()
+            if file_name:
+                with open(file_name, "wb") as output:
+                    num_bytes = output.write(figure_data)
```

Comment:
Again, `figure_data` can be matplotlib `Figure`. I tested this code locally but 
```
TypeError                                 Traceback (most recent call last)
<ipython-input-25-b61895a17477> in <module>
      1 with open('test_data', 'wb') as fp:
----> 2     datasize = fp.write(fig)

TypeError: a bytes-like object is required, not 'Figure'
```

Reply:
I'm not sure how to handle `bytes` from Jessies PR, so I guess i should remove it and just use savefig from matplotlib for now